src/main/java/org/hydev/mcpm/client/interaction/ProgressBar.java

```diff
@@ -18,50 +34,61 @@ public class ProgressBar implements AutoCloseable
     public ProgressBar(ProgressBarTheme theme)
     {
         this.theme = theme;
-        this.init();
-    }
+        this.out = System.out;
+        this.cu = new ConsoleUtils(this.out);
+        this.activeBars = new ArrayList<>();
+        this.cols = AnsiConsole.getTerminalWidth();
 
-    /**
-     * Initialize the progress bar (print the first line)
-     */
-    public void init()
-    {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        // Last update time
+        this.lastUpdate = System.nanoTime();
     }
 
     /**
      * Append a progress bar at the end
      *
-     * @return Unique identifier of the progress bar
+     * @param bar Row of the progress bar
+     * @return bar for fluent access
      */
-    public String appendBar()
+    public ProgressRow appendBar(ProgressRow bar)
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        this.activeBars.add(bar);
+        bar.setPb(this);
+
+        out.println();
+        update();
+        return bar;
     }
 
-    /**
-     * Set progress for a bar
-     *
-     * @param id Unique identifier
-     * @param progress Progress as a ratio in range 0-1
-     */
-    public void setBar(String id, float progress)
+    protected void update()
     {
-        // TODO: Implement this
-        throw new UnsupportedOperationException("TODO");
+        // Check time, update only every 0.0166s (60 fps)
+        // Performance of the update heavily depends on the terminal's escape code handling
+        // implementation, so frequent updates will degrade performance on a bad terminal
+        var curTime = System.nanoTime();
+        if ((curTime - lastUpdate) / 1e9d < 0.0166) return;
```

Comment:
Magic numbers? Make these constants or take them in as a parameter.

Reply:
It's written in the comment above, that 0.0166 is the delay between each frame for 60fps. And `1e9` is pretty straightforward for converting nanoseconds to seconds.

Reply:
I guess I can make fps configurable

Reply:
This is actually an anti-pattern IIRC. Comments help but its wayy better to give them a formal name and move them out the code "battle zone."